File: I/350-intersection-of-two-arrays-ii/main.go
package main

import (
	"fmt"
)

func intersect(nums1 []int, nums2 []int) []int {

	var size int
	var t1, t2 []int
	if len(nums1) > len(nums2) {
		size = len(nums2)
		t1 = nums2
		t2 = nums1
	} else {
		size = len(nums1)
		t1 = nums1
		t2 = nums2
	}
	ret := make([]int, size)

	r1 := 0
	for i := 0; i < len(t1); i++ {
		for j := 0; j < len(t2); j++ {
			if t1[i] == t2[j] {
				ret[r1] = t1[i]
				r1++
				t := make([]int , len(t2)-1)
				copy(t,t2[0:j])
				copy(t[j:len(t)],t2[j+1:len(t2)])
				t2 = t
				break
			}
		}
	}
	r := make([]int ,r1)
	copy(r, ret[0:r1])
	return r
}

func main() {
	fmt.Println(intersect([]int{1, 2, 2, 1}, []int{2, 2}))
	fmt.Println(intersect([]int{4,9,5}, []int{9,4,9,8,4}))
	fmt.Println(intersect([]int{3,1,2}, []int{1,1}))
}

File: I/350-intersection-of-two-arrays-ii/main.rs
struct Solution {}

use std::collections::HashMap;
impl Solution {
  pub fn intersect(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
    let mut m: HashMap<i32, i32> = HashMap::new();

    let s: &Vec<i32> = if nums1.len() > nums2.len() {
      nums2.as_ref()
    } else {
      nums1.as_ref()
    };

    let l: &Vec<i32> = if nums1.len() > nums2.len() {
      nums1.as_ref()
    } else {
      nums2.as_ref()
    };

    s.iter().for_each(|&v| {
      m.entry(v).and_modify(|x| *x += 1).or_insert(1);
    });
    let mut ans: Vec<i32> = Vec::new();
    l.iter().for_each(|&v| {
      m.entry(v).and_modify(|x| {
        if *x > 0 {
          ans.push(v);
        }
        *x -= 1;
      });
    });
    ans
  }
}

fn main() {
  println!("{:?}", Solution::intersect(vec![1, 2, 3], vec![2, 3, 4]));
}
